File: build_project/compile_resource_pack.py
#!/usr/bin/env python3
import os
import sys
from . import utils
from . import create_texture
from . import png_to_opa

TEMPLATE_PREFIX = "Template_"


def compile_resource_pack(input_resource_pack, output_resource_pack):
    if not input_resource_pack.endswith("/"):
        input_resource_pack += "/"
    if not output_resource_pack.endswith("/"):
        output_resource_pack += "/"

    utils.createDir(output_resource_pack)
    for file in os.listdir(input_resource_pack):
        input_file = input_resource_pack + file

        if file == ".DS_Store":
            continue

        if file.endswith(".png"):
            if file.startswith(TEMPLATE_PREFIX):
                output_file = output_resource_pack + file[len(TEMPLATE_PREFIX):-4] + ".opa"
            else:
                output_file = output_resource_pack + file[0:-4] + ".opa"
        else:
            output_file = output_resource_pack + file

        if os.path.exists(output_file) and not os.path.isdir(output_file) and os.path.getmtime(input_file) <= os.path.getmtime(output_file):
            continue

        if os.path.isdir(input_file):
            compile_resource_pack(input_file, output_file)
        else:
            if file.endswith(".png"):
                if file.startswith(TEMPLATE_PREFIX):
                    temp_file = "._temp.png"
                    create_texture.generate_block_texture(input_file, temp_file)
                    png_to_opa.png_to_opa(temp_file, output_file)
                    utils.remove(temp_file)
                else:
                    png_to_opa.png_to_opa(input_file, output_file)
            else:
                utils.copy(input_file, output_file)


if __name__ == "__main__":
    if len(sys.argv) != 3:
        print("Usage: compileResourcePack.py [input] [output]")
    else:
        compileResourcePack(sys.argv[1], sys.argv[2])

File: build_project/compile_resource_pack.rs
use std::path::PathBuf;

use crate::build_project::png_to_opa::png_file_to_opa_file;

pub fn compile_resource_pack(input_resource_pack: PathBuf, output_resource_pack: PathBuf) {
    // make sure that cargo reruns this script if the input resource pack changes
    println!(
        "cargo:rerun-if-changed={}",
        input_resource_pack.to_str().unwrap()
    );
    // also rerun if the output resource pack changes
    println!(
        "cargo:rerun-if-changed={}",
        output_resource_pack.to_str().unwrap()
    );

    // create output_resource_pack directory if it doesn't exist
    std::fs::create_dir_all(output_resource_pack.clone()).unwrap();

    // iterate through all files in input_resource_pack
    for entry in std::fs::read_dir(input_resource_pack).unwrap() {
        let path = entry.unwrap().path();

        // skip .DS_Store files
        if path.file_name().unwrap().to_str().unwrap() == ".DS_Store" {
            continue;
        }
        // if its a folder then recursively call this function
        else if path.is_dir() {
            compile_resource_pack(
                path.clone(),
                output_resource_pack.join(path.file_name().unwrap()),
            );
        }
        // if the file ends with .png then convert it to .opa
        else if path.extension().unwrap() == "png" {
            png_file_to_opa_file(
                path.clone(),
                output_resource_pack.join(
                    path.file_name()
                        .unwrap()
                        .to_str()
                        .unwrap()
                        .replace(".png", ".opa"),
                ),
            );
        }
        // else just copy the file
        else {
            std::fs::copy(
                path.clone(),
                output_resource_pack.join(path.file_name().unwrap()),
            )
                .unwrap();
        }
    }
}
